linux-sandbox.jl/mk_cgroup.c

```diff
@@ -0,0 +1,180 @@
+    // clobbers its input, so we copy to a temporary variable first. >:|
+    char dir_dirname[PATH_MAX];
+    strncpy(dir_dirname, dir, PATH_MAX - 1);
+    mkpath(dirname(&dir_dirname[0]));
+
+    // then create our directory
+    int result = mkdir(dir, 0777);
+    check((0 == result));
+}
+
+static int chown_r(const char * path, uid_t uid, gid_t gid) {
+    DIR * dir = opendir(path);
+    if (!dir) {
+        return 1;
+    }
+    check(chown(path, uid, gid) == 0);
+
+    struct dirent *entry;
+    char leaf_path[PATH_MAX];
+    while ((entry = readdir(dir)) != NULL) {
+        if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0)
+            continue;
+
+        snprintf(leaf_path, sizeof(leaf_path), "%s/%s", path, entry->d_name);
+        check(chown(leaf_path, uid, gid) == 0);
+
+        if (entry->d_type == DT_DIR) {
+            if (chown_r(leaf_path, uid, gid) != 0) {
+                return 1;
+            }
+        }
+    }
+    closedir(dir);
+
+    return 0;
+}
+
+static int verify_cgroup(const char * name, uid_t uid, gid_t gid, const char * cpus) {
+    // First, check if the directory exists:
+    char path[PATH_MAX];
+    snprintf(path, sizeof(path), "/sys/fs/cgroup/cpuset/%s", name);
+    struct stat sb;
+    if (stat(path, &sb) != 0 || !S_ISDIR(sb.st_mode)) {
+        return 1;
+    }
+
+    // Ensure it's owned by the right uid and gid
+    if (sb.st_uid != uid || sb.st_gid != gid) {
+        return 1;
+    }
+
+    // Ensure the assigned cpus are correct
+    snprintf(path, sizeof(path), "/sys/fs/cgroup/cpuset/%s/cpuset.cpus", name);
+    int fd = open(path, O_RDONLY);
+    if (fd == -1) {
+        return 1;
+    }
+    char buff[128];
+    int bytes_read = read(fd, buff, sizeof(buff));
+    close(fd);
+    if (bytes_read <= 0) {
+        return 1;
+    }
+    if (strncmp(cpus, buff, min(bytes_read, strlen(cpus))) != 0) {
+        return 1;
+    }
+
+    // Ensure the assigned mems are correct
+    snprintf(path, sizeof(path), "/sys/fs/cgroup/cpuset/%s/cpuset.mems", name);
+    fd = open(path, O_RDONLY);
+    if (fd == -1) {
+        return 1;
+    }
+    bytes_read = read(fd, buff, sizeof(buff));
+    close(fd);
+    if (bytes_read <= 0) {
+        return 1;
+    }
+
+    snprintf(path, sizeof(path), "/sys/fs/cgroup/cpuset/cpuset.mems");
+    fd = open(path, O_RDONLY);
+    if (fd == -1) {
+        return 1;
+    }
+    char true_mems[128];
+    bytes_read = read(fd, true_mems, sizeof(true_mems));
+    if (bytes_read <= 0) {
+        return 1;
+    }
+    if (strncmp(true_mems, buff, bytes_read) != 0) {
+        return 1;
+    }
+
+    // If all of those tests passed, the cgroup is setup correctly!
+    return 0;
+}
+
+int main(int argc, char * argv[]) {
+    if( argc < 5 ) {
+        fprintf(stderr, "Usage: mk_cgroup <name> <uid> <gid> <cpus>\n");
+        return 1;
+    }
+    const char * name = argv[1];
+    uid_t uid = atoi(argv[2]);
```

Comment:
Should probably pick up the users uid/gid rather than taking them as arguments. (When doing setuid, the original uid/gid is remembered in of getuid/geteuid, I forget the exact details, but it'll be in the man pages).